ballerina-shell/modules/shell-core/src/main/java/io/ballerina/shell/invoker/classload/ImportsManager.java

```diff
@@ -0,0 +1,314 @@
+        storeImport(quotedPrefix, importedModule);
+    }
+
+    /**
+     * Add prefixes to persisted list of imports.
+     * The name will be linked with the import prefix.
+     *
+     * @param name     Usage source declaration name of the import.
+     * @param prefixes Used prefixes.
+     */
+    public void storeImportUsages(QuotedIdentifier name, Collection<QuotedIdentifier> prefixes) {
+        // Get the prefixes previously used by this name and add prefix this to it.
+        Set<QuotedIdentifier> sourcePrefixes = this.usedPrefixes.getOrDefault(name, new HashSet<>());
+        sourcePrefixes.addAll(prefixes);
+        this.usedPrefixes.put(name, sourcePrefixes);
+    }
+
+    /**
+     * All the prefixes that were added. Prefixes will be quoted.
+     * These are essentially all the imports that are managed.
+     *
+     * @return Set of prefixes.
+     */
+    public Set<QuotedIdentifier> prefixes() {
+        return this.imports.keySet();
+    }
+
+    /**
+     * All the import statements that were used by the given names.
+     * This will also contain all the default imports. eg: ballerina/jballerina.java
+     * Names should be quoted identifiers of module dclns and var dclns.
+     *
+     * @param names Names to search the used prefixes for.
+     * @return Set of import statements used in given names.
+     */
+    public Set<String> getUsedImports(Collection<QuotedIdentifier> names) {
+        // Start with all default imports
+        Set<QuotedIdentifier> allUsedImportPrefixes = new HashSet<>(usedPrefixes.get(ANON_SOURCE));
+
+        // Add the imports used by names
+        names.stream()
+                .map(this.usedPrefixes::get) // get used prefixes of names
+                .filter(Objects::nonNull) // discard null lists
+                .forEach(allUsedImportPrefixes::addAll);
+
+        // Compose import strings
+        Set<String> importStrings = new HashSet<>();
+        allUsedImportPrefixes.stream()
+                .map(this::getImport) // get the import statement
+                .filter(Objects::nonNull) // discard null imports
+                .forEach(importStrings::add);
+
+        return importStrings;
+    }
+
+    /**
+     * Converts a type symbol into its string counterpart.
+     * Any imports that need to be done are calculated.
+     * Eg: if the type was abc/z:TypeA then it will be converted as
+     * 'import abc/z' will be added as an import.
+     * We need to traverse all the sub-typed because any sub-type
+     * may need to be imported.
+     * Also, this will compute all the replacements that should be done.
+     * In above case, abc/z:TypeA will be replaced with z:TypeA.
+     *
+     * @param typeSymbol Type to process.
+     * @param imports    Imports set to store imports.
+     * @return Type signature after parsing.
+     */
+    protected String extractImportsFromType(TypeSymbol typeSymbol, Set<QuotedIdentifier> imports) {
```

Comment:
IIUC, the purpose of this method is to figure out which module to import from the given type symbol right? Can't the module symbol returned from the `getModule()` method be used for that?

Reply:
The purpose is to figure out the required imports to use a type. so the type is something like (xyz/pqr:C | xyz/abc:D) and from this we have to extract imports (eg: import xyz/pqr, import xyz/abc) and replace signature as (pqr:C | abc:D)

This is the same approach taken in lang server iirc

Reply:
Still for a union type symbol, it'd be easier to derive the import statements using the module symbols of the member types of the union though.

Reply:
I'll create an issue for this to track this. #29109